### PR TITLE
Add custom Jest matchers for AMP validation

### PR DESCRIPTION
**Pull request by swissspidy**
Adds the following new Jest matchers:

* `toBeValidAMP`
* `toBeValidAMPStoryPage` - like `toBeValidAMP`, but wraps the component/string in `<amp-story>` first
* `toBeValidAMPStoryElement` - like `toBeValidAMPStoryPage`, but wraps the component/string in `<amp-story-page>` and `<amp-story-grid-layer>` first

Usage:

```js
await expect(<VideoOutput {...props} />).toBeValidAMPStoryElement();
```

To test:

1. Apply the following patch:

```diff
diff --git assets/src/edit-story/elements/video/output.js assets/src/edit-story/elements/video/output.js
index d7d7c784..54255cb7 100644
--- assets/src/edit-story/elements/video/output.js
+++ assets/src/edit-story/elements/video/output.js
@@ -52,7 +52,7 @@ function VideoOutput({
     src,
   };
   const props = {
-    autoPlay: 'autoplay',
+    autoPlay: true,
     poster,
     layout: 'fill',
   };
```

2. Run `npm run test:js assets/src/edit-story/elements/video/test/output.js`
3. See the following output:

```
Video output › should produce valid AMP output

Expected <div style="position:absolute;width:177.77778%;height:100%;left:-38.88889%;top:0%"><amp-video autoPlay="true" poster="https://example.com/poster.png" layout="fill"><source type="video/mp4" src="https://example.com"/></amp-video></div> to be valid AMP. Errors:
The attribute 'autoplay' in tag 'amp-story >> amp-video' is set to the invalid value 'true'. (see https://amp.dev/documentation/components/amp-video)
```

This exact error was actually caught only because of the work on #346. I figured adding some tests for our AMP validity would make sense.

While it's not exactly how AMP validation works, and is not perfect (lots of assumptions right now), it seems to do the trick.

I also considered running the validator every time a story is saved, but `amphtml-validator` doesn't run in the browser.

### CODE REVIEW

**Comment by swissspidy** [comment]
Could perhaps also use AMP Optimizer:

https://github.com/ampproject/amp-toolbox/tree/master/packages/optimizer#incomplete-markup

File: assets/src/edit-story/elements/video/test/output.js [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+          videoId: 123,
+          src: 'https://example.com/image.png',
+          height: 1920,
+          width: 1080,
+        },
+      },
+      box: { width: 1080, height: 1920, x: 50, y: 100, rotationAngle: 0 },
+    };
+
+    await expect(<VideoOutput {...props} />).toBeValidAMP();
```
Comment (dvoytenko) [inline assets/src/edit-story/elements/video/test/output.js]: Trying to figure out where in this system the `<script custom-element="amp-video">` is inserted before testing.

**Review by dvoytenko — DISMISSED** [review]
This looks awesome! Just one question.

File: assets/src/edit-story/elements/video/test/output.js (hunk 1 above)
Comment (swissspidy) [reply assets/src/edit-story/elements/video/test/output.js]: The AMP Optimizer in the matcher would add any missing extensions.

File: tests/js/matchers/toBeValidAMP.js [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+ *
+ * @see https://github.com/ampproject/amp-toolbox/tree/master/packages/optimizer#incomplete-markup
+ *
+ * @todo Add option to disable AMP Optimizer?
+ *
+ * @param {string|Element} stringOrComponent String or component to test.
+ * @return {Object} returns Jest matcher object
+ */
+async function toBeValidAMP(stringOrComponent) {
+  const string = renderToStaticMarkup(stringOrComponent);
```
Comment (swissspidy) [inline tests/js/matchers/toBeValidAMP.js:28]: Presumably **all** elements need to be wrapped in `amp-story` / `amp-story-page` / `amp-story-grid-layer` here for the proper context. An `amp-video` in a regular AMP page is different from one in a story. Just like `amp-youtube` is allowed in a regular AMP page but not in a story.

**Review by wassgha — APPROVED** [review]
Love this!

File: tests/js/matchers/toBeValidAMP.js [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+/**
+ * Jest matcher for asserting that a string or component is valid AMP.
+ *
+ * Uses AMP Optimizer to add any missing tags and extension imports.
+ *
+ * @see https://github.com/ampproject/amp-toolbox/tree/master/packages/optimizer#incomplete-markup
+ *
+ * @todo Add option to disable AMP Optimizer?
```
Comment (swissspidy) [inline tests/js/matchers/toBeValidAMP.js]: Todo: allow passing options to Optimizer?

File: tests/js/matchers/toBeValidAMP.js [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+ * limitations under the License.
+ */
+
+/**
+ * External dependencies
+ */
+import { renderToStaticMarkup } from 'react-dom/server';
+
+/**
+ * Internal dependencies
```
Comment (swissspidy) [inline tests/js/matchers/toBeValidAMP.js:23]: Wrong docblock

File: tests/js/matchers/toBeValidAMP.js (hunk 2 above)
Comment (swissspidy) [reply tests/js/matchers/toBeValidAMP.js:28]: Done now by partially reverting the previous change

File: assets/src/edit-story/elements/video/test/output.js (hunk 1 above)
Comment (swissspidy) [reply assets/src/edit-story/elements/video/test/output.js]: Reflected this now in the docblock for `getAMPValidationErrors`

**Comment by swissspidy** [comment]
Just noticed that `WithElementMask` doesn't produce valid AMP, see d0bdd5e0d6b1a3acfb59c1e4a747ad8f219ba3f5.

> The property 'clip-path' in attribute 'style' in tag 'div' is disallowed. (see https://amp.dev/documentation/guides-and-tutorials/develop/style_and_layout/style_pages)

Looks like AMP only allows it on `svg` elements.

@dvoytenko Any thoughts on this?

**Comment by dvoytenko** [comment]
> Just noticed that `WithElementMask` doesn't produce valid AMP, see d0bdd5e0d6b1a3acfb59c1e4a747ad8f219ba3f5.

I don't see this documented anywhere in AMP docs. So, we should definitely reach out to them to confirm.

In the meantime, we need to work it around (@wassgha PTAL). We could do something like this:
 1. Scan all elements when building `<style amp-custom>` (#346 enables this and already merged). For each unique mask we add `.mask-<type> {clip-path: url(#mask-<type>)}` and whatever else might be needed.
 2. Somewhere convenient in DOM, higher up, we add `<div [move off screen]>...</div>`. For each unique mask we add `<svg id="mask-<type>">...</svg>`.

@swissspidy @wassgha WDYT?

**Comment by swissspidy** [comment]
I didn’t see it explicitly in the docs, but the protoascii spec file defines it that way.

The workaround sounds reasonable though.

**Comment by dvoytenko** [comment]
Let's wait and see if we can get https://github.com/ampproject/amphtml/pull/27160 through. If not - we'll just workaround on our side.

**Comment by dvoytenko** [comment]
To note though, there are some semantical benefits of doing what I described [above](https://github.com/google/web-stories-wp/pull/500#issuecomment-596745077) - it'd be structurally more appropriate approach. But it will also take more time, so I'd rather see what we have to deal with first.

**Comment by swissspidy** [comment]
Sounds good to me. We can create a new issue/PR for that change if needed.

**Pull request merged**